doc/stp/SONiC_PVST_HLD.md

```diff
@@ -0,0 +1,700 @@
+The scaling limit might differ depending on the platform and the CPU used, which needs to be determined based on testing.
+
+## 1.4 Warm Boot Requirements
+
+Warm boot will not be supported
+
+# 2 Functionality
+
+## 2.1 Functional Description
+The Spanning Tree Protocol (STP) prevents Layer 2 loops in a network and provides redundant links. If a primary link fails, the backup link is activated and network traffic is not affected. STP also ensures that the least cost path is taken when multiple paths exist between the devices.
+
+When the spanning tree algorithm is run, the network switches transform the real network topology into a spanning tree topology. In an STP topology any LAN in the network can be reached from any other LAN through a unique path. The network switches recalculate a new spanning tree topology whenever there is a change to the network topology.
+
+For each switch in the topology, a port with lowest path cost to the root bridge is elected as root port.
+
+For each LAN, the switches that attach to the LAN select a designated switch that is the closest to the root switch. The designated switch forwards all traffic to and from the LAN. The port on the designated switch that connects to the LAN is called the designated port. The switches decide which of their ports is part of the spanning tree. A port is included in the spanning tree if it is a root port or a designated port.
+
+PVST+ allows for running multiple instances of spanning tree on per VLAN basis. 
+
+One of the advantage with PVST is it allows for load-balancing of the traffic. When a single instance of spanning tree is run and a link is put into blocking state for avoiding the loop, it will result in inefficient bandwidth usage. With per VLAN spanning tree multiple instances can be run such that for some of the instances traffic is blocked over the link and for other instances traffic is forwared allowing for load balancing of traffic.
+ 
+PVST+ support allows the device to interoperate with IEEE STP and also tunnel the PVST+ BPDUs transparently across IEEE STP region to potentially connect other PVST+ switches across the IEEE STP region. For interop with IEEE STP, PVST+ will send untagged IEEE BPDUs (MAC - 01:80:C2:00:00:00) with information corresponding to VLAN 1. The STP port must be a member of VLAN 1 for interoperating with IEEE STP.
+
+# 3 Design
+## 3.1 Overview
+
+![STP](images/STP_Architecture.png "Figure 1: High level architecture")
+__Figure 1: High level architecture__
+
+High level overview:
+
+STP container will host STPMgr and STPd process.
+
+STPMgr will handle all the STP configurations and VLAN configurations via config DB and state DB respectively.
+
+STPd process will handle all the protocol functionality and has following interactions
+
+ * Linux kernel for packet tx/rx and netlink events
+ * STPMgr interaction for configuration handling
+ * STPSync is part of STPd handling all the STP operational data updates to APP DB 
```

Comment:
it looks like the kernel stp state is not programmed at all, there will be inconsistency between the asic and config db. is there going to be any case that such inconsistency will break?

Reply:
Thanks for bringing this point !! We did some investigation on this and with the vlan aware bridge model being used in SONiC we don’t have a way to program the STP state, it’s possible in case of linux traditional bridge where a netdevice is created for each vlan, port combination by updating the netdevice state. 

Below are few alternatives to achieve the STP blocking functionality in kernel to keep it in sync with HW state, request your feedback on this -
1) When STP state is blocking remove the VLAN membership on the port from the vlan aware bridge, this will ensure the traffic is not processed and L2 control packets like STP can still be trapped from kernel to user space as they rely on PF_PACKET socket. 

2) Use linux tc (traffic control) filtering mechanism where we could drop specific VLAN packets on an interface. This may not be efficient compared to 1) as we may need to program more number of filters on each device(or interface) on per VLAN basis.

Reply:
Updated the document